src/ddsrt/src/sync/freertos/sync.c

```diff
@@ -0,0 +1,505 @@
+{
+  assert(rwlock != NULL);
+  ddsrt_mutex_lock(&rwlock->mutex);
+  while (rwlock->wtrs > 0) {
+    ddsrt_cond_wait(&rwlock->cond, &rwlock->mutex);
+  }
+  assert(rwlock->state == UNLOCKED ||
+         rwlock->state == LOCKED_SHARED);
+  rwlock->wtrs++;
+  while (rwlock->rdrs > 0) {
+    ddsrt_cond_wait(&rwlock->cond, &rwlock->mutex);
+  }
+  assert(rwlock->state == UNLOCKED);
+  rwlock->state = LOCKED_EXCLUSIVE;
+  ddsrt_mutex_unlock(&rwlock->mutex);
+}
+
+bool ddsrt_rwlock_trywrite(ddsrt_rwlock_t *rwlock)
+{
+  bool locked = false;
+  assert(rwlock != NULL);
+  ddsrt_mutex_lock(&rwlock->mutex);
+  if (rwlock->wtrs == 0 && rwlock->rdrs == 0) {
+    assert(rwlock->state == UNLOCKED);
+    rwlock->wtrs++;
+    rwlock->state = LOCKED_EXCLUSIVE;
+    locked = true;
+  }
+  ddsrt_mutex_unlock(&rwlock->mutex);
+  return locked;
+}
+
+void ddsrt_rwlock_unlock(ddsrt_rwlock_t *rwlock)
+{
+  assert(rwlock != NULL);
+  ddsrt_mutex_lock(&rwlock->mutex);
+  assert(rwlock->state != UNLOCKED);
+  if (rwlock->state == LOCKED_EXCLUSIVE) {
+    assert(rwlock->rdrs == 0);
+    assert(rwlock->wtrs > 0);
+    rwlock->wtrs--;
+    rwlock->state = UNLOCKED;
+  } else {
+    assert(rwlock->rdrs > 0);
+    rwlock->rdrs--;
+    if (rwlock->rdrs == 0) {
+      rwlock->state = UNLOCKED;
+    }
+  }
+  if (rwlock->state == UNLOCKED) {
+    ddsrt_cond_broadcast(&rwlock->cond);
+  }
+  ddsrt_mutex_unlock(&rwlock->mutex);
+}
+
+#define ONCE_NOT_STARTED (1<<0)
+#define ONCE_IN_PROGRESS (1<<1)
+#define ONCE_FINISHED (1<<2)
+
+/* Wait one millisecond (tick) between polls. */
+static const TickType_t once_delay = (configTICK_RATE_HZ / 1000);
+
+void
+ddsrt_once(
+  ddsrt_once_t *control,
+  ddsrt_once_fn init_fn)
+{
+  int ret, brk = 0;
+  uint32_t stat;
+
+  while (brk == 0) {
+    stat = ddsrt_atomic_ld32(control);
+    /* Verify once control was initialized properly. */
+    assert(stat == ONCE_NOT_STARTED ||
+           stat == ONCE_IN_PROGRESS ||
+           stat == ONCE_FINISHED);
+
+    if ((stat & ONCE_FINISHED) != 0) {
+      /* The initialization function has been executed. No reason to block
+         execution of this thread. Continue. */
+      brk = 1;
+    } else if ((stat & ONCE_IN_PROGRESS) != 0) {
+      /* Another thread is executing the initialization function. Wait around
+         for it to be finished. The polling loop is required because FreeRTOS
+         does not offer futexes. */
+      vTaskDelay(once_delay);
+      /* Repeat. */
+    } else {
+      /* No thread was executing the initialization function (one might be
+         executing it now) at the time of the load. If the atomic compare and
+         swap operation is successful, this thread will run the initialization
+         function. */
+      if (ddsrt_atomic_cas32(
+            control, ONCE_NOT_STARTED, ONCE_IN_PROGRESS) != 0)
+      {
+        /* Function must never block or yield, see reference manual. */
+        init_fn();
+
+        ret = (0 == ddsrt_atomic_cas32(
```

Comment:
cas returns false must be false ... ahum ... there's an easier way to write that!

You could also use a bitmask instead and use atomic_or32_ov instead of calling cas32